### PR TITLE
staticjinja: enable tests; lift to top level

### PR DESCRIPTION
**Pull request by fgaz**
###### Motivation for this change

Lifted to the top level because most users will probably use the cli tool

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/staticjinja/default.nix [hunk 1]
```diff
@@ -8,17 +8,18 @@
...
 
-  propagatedBuildInputs = [ jinja2 docopt easywatch ];
-
-  # There are no tests on pypi
-  doCheck = false;
+  propagatedBuildInputs = [
+    jinja2
+    docopt
+    easywatch
+  ];
```
Comment (jonringer) [inline pkgs/development/python-modules/staticjinja/default.nix:22]: If tests are not available, then please use `pythonImportsCheck` to import the most important modules. This isn't as good as unit tests, but will usually give a good indication of run-time errors. Please see [pythonImportsCheck documentation](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#using-pythonimportscheck) for more information.

Comment (fgaz) [reply pkgs/development/python-modules/staticjinja/default.nix:31]: I managed to fix the unit tests! The trick was using the source from github _and_ ignoring path mismatches

File: pkgs/development/python-modules/staticjinja/default.nix [hunk 2]
```diff
@@ -1,24 +1,43 @@
...
-    inherit pname version;
-    sha256 = "fbd61cca1dad44b6891d1a1d72b11ae100e21b3909802e3ff1861ab55bf16603";
+  disabled = isPy27; # 0.4.0 drops python2 support
+
+  # For some reason, in pypi the tests get disabled when using
+  # PY_IGNORE_IMPORTMISMATCH, so we just fetch from GitHub
+  src = fetchFromGitHub {
+    owner = "staticjinja";
+    repo = pname;
+    rev = "0.4.0";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/staticjinja/default.nix:23]:
```suggestion
    rev = version;
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 104401` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>python37Packages.staticjinja</li>
    <li>staticjinja (python38Packages.staticjinja)</li>
  </ul>
</details>

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 104401` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>python37Packages.staticjinja</li>
    <li>staticjinja (python38Packages.staticjinja)</li>
  </ul>
</details>

**Comment by FRidh** [comment]
> Lifted to the top level because most users will probably use the cli tool

however, the tool will import the Python code one wrote which imports other modules. 

I think its a bad idea to expose this one as top-level attribute and think that it should always be considered as part of a Python environment that has all the modules that one uses in a staticjinja project. Or are you saying one *never* uses additional Python libraries with staticjinja?

**Comment by fgaz** [comment]
@FRidh I'm saying there's two ways of using this library/program

1. By [using it as a dependency](https://staticjinja.readthedocs.io/en/latest/user/advanced.html) in a python environment.
2. By just using [the command line tool](https://staticjinja.readthedocs.io/en/latest/user/quickstart.html#rendering-templates) (`staticjinja build`), like in most site generators. This might as well be in a larger python project, but in this case staticjinja is treated only as an executable, not as a library. (this is the only way I personally used this project)

Having the executable at the top level facilitates the second, simpler, way

File: pkgs/development/python-modules/staticjinja/default.nix (hunk 2 above)
Comment (fgaz) [reply pkgs/development/python-modules/staticjinja/default.nix]: Thanks. Done

**Comment by FRidh** [comment]
> By just using the command line tool (staticjinja build), like in most site generators. This might as well be in a larger python project, but in this case staticjinja is treated only as an executable, not as a library. (this is the only way I personally used this project)

I understand, but would you argue that one typically does so without any additional Python libraries being used aside from those required by staticjinja?

**Comment by fgaz** [comment]
I'm not sure I understand, could you make an example of using some python library together with staticjinja (the executable), and having the second interfere with the first?

I typically just do something like this (taken from the README):

```
$ mkdir templates
$ vim templates/index.html
$ staticjinja watch
Building index.html...
Templates built.
Watching 'templates' for changes...
Press Ctrl+C to stop.
```

Basically I just use it as a simple way to run a template engine. As far as I can see there isn't even a way to use additional python libraries when running it like this.

**Comment by FRidh** [comment]
Okay, I see. I mostly skimmed through the advanced section, thinking that becomes pretty essential considering the basic functionality is so limited.

**Comment by fgaz** [comment]
It depends. If you don't have to load external data (ex. for building simple sites, or for prototyping or reworking a bunch of existing pages) it's surprisingly powerful. If you prefer I can remove the second commit though.

**Comment by fgaz** [comment]
Ping

**Comment by fgaz** [comment]
Oh wait, I see that @FRidh updated it independently in https://github.com/NixOS/nixpkgs/commit/722c434a18776db62ed062c686ee8fbc1bccaed9. Let me rebase...

File: pkgs/development/python-modules/staticjinja/default.nix [hunk 3]
```diff
@@ -1,24 +1,43 @@
...
 
-  propagatedBuildInputs = [ jinja2 docopt easywatch ];
+  propagatedBuildInputs = [
+    jinja2
+    docopt
+    easywatch
+  ];
+
+  checkInputs = [
+    pytest
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/staticjinja/default.nix:34]:
```suggestion
    pytestCheckHook
```

File: pkgs/development/python-modules/staticjinja/default.nix [hunk 4]
```diff
@@ -1,24 +1,43 @@
 { lib
-, fetchPypi
+, fetchFromGitHub
 , buildPythonPackage
+, isPy27
 , docopt
 , easywatch
 , jinja2
+, pytest
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/staticjinja/default.nix:8]:
```suggestion
, pytestCheckHook
```

File: pkgs/development/python-modules/staticjinja/default.nix [hunk 5]
```diff
@@ -1,24 +1,43 @@
...
+  checkInputs = [
+    pytest
+    markdown
+  ];
 
-  # There are no tests on pypi
-  doCheck = false;
+  # Import paths differ by a "build/lib" subdirectory, but the files are
+  # the same, so we ignore import mismatches.
+  checkPhase = "PY_IGNORE_IMPORTMISMATCH=1 pytest";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/staticjinja/default.nix:40]:
```suggestion
  precheck = ''
    export PY_IGNORE_IMPORTMISMATCH=1
  '';
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/staticjinja/default.nix (hunk 3 above)
Comment (fgaz) [reply pkgs/development/python-modules/staticjinja/default.nix]: Done (and the rest too)

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review which does not build all packages (e.g. lumo, tensorflow or pytorch)
If you find some bugs or got suggestions for further things to search or run please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 104401` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
3 packages built:
</summary>
<ul>
<li>python37Packages.staticjinja</li>
<li>staticjinja (python38Packages.staticjinja)</li>
<li>python39Packages.staticjinja</li>
</ul>
</details>

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review which does not build all packages (e.g. lumo, tensorflow or pytorch)
If you find some bugs or got suggestions for further things to search or run please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 104401` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
3 packages built:
</summary>
<ul>
<li>python37Packages.staticjinja</li>
<li>staticjinja (python38Packages.staticjinja)</li>
<li>python39Packages.staticjinja</li>
</ul>
</details>

**Pull request merged**